envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/ServiceTagsAndCanaryTest.kt

```diff
@@ -32,134 +25,11 @@ class ServiceTagsAndCanaryTest {
-
-        // when
-        val stats = callEchoServiceRepeatedly(repeat = 10, tag = "ipsum", canary = true)
+    override fun consul() = consul
 
-        // then
-        assertThat(stats.totalHits).isEqualTo(10)
-        assertThat(stats.loremCanaryHits).isEqualTo(0)
-        assertThat(stats.loremRegularHits).isEqualTo(0)
-        assertThat(stats.ipsumRegularHits).isEqualTo(10)
-    }
-
-    @Test
-    fun `should route requests to canary instance`() {
-        // given
-        registerServices()
-        waitForReadyServices("echo")
-
-        // when
-        val stats = callEchoServiceRepeatedly(repeat = 10, canary = true)
-
-        // then
-        assertThat(stats.totalHits).isEqualTo(10)
-        assertThat(stats.loremCanaryHits).isEqualTo(10)
-        assertThat(stats.loremRegularHits).isEqualTo(0)
-        assertThat(stats.ipsumRegularHits).isEqualTo(0)
-    }
-
-    @Test
-    fun `should return 503 if no instance with requested tag is found`() {
-        // given
-        registerServices()
-        waitForReadyServices("echo")
-
-        // when
-        val stats = callEchoServiceRepeatedly(repeat = 10, tag = "not-found", canary = true, assertNoErrors = false)
-
-        // then
-        assertThat(stats.totalHits).isEqualTo(10)
-        assertThat(stats.failedHits).isEqualTo(10)
-        assertThat(stats.loremCanaryHits).isEqualTo(0)
-        assertThat(stats.loremRegularHits).isEqualTo(0)
-        assertThat(stats.ipsumRegularHits).isEqualTo(0)
-    }
-
-    fun waitForReadyServices(vararg serviceNames: String) {
-        serviceNames.forEach {
-            untilAsserted {
-                envoy.egressOperations.callService(it).also {
-                    assertThat(it).isOk()
-                }
-            }
-        }
-    }
-
-    fun callEchoServiceRepeatedly(
-        repeat: Int,
-        tag: String? = null,
-        canary: Boolean,
-        assertNoErrors: Boolean = true
-    ): CallStats {
-        val stats = CallStats(listOf(
-                loremCanaryService.container, loremRegularService.container, ipsumRegularService.container
-        ))
-        val tagHeader = tag?.let { mapOf("x-service-tag" to it) } ?: emptyMap()
-        val canaryHeader = if (canary) mapOf("x-canary" to "1") else emptyMap()
-
-        envoy.egressOperations.callServiceRepeatedly(
-            service = "echo",
-            stats = stats,
-            minRepeat = repeat,
-            maxRepeat = repeat,
-            headers = tagHeader + canaryHeader,
-            assertNoErrors = assertNoErrors
-        )
-        return stats
-    }
+    override fun envoyControl() = envoyControl
 
-    val CallStats.loremCanaryHits: Int
-        get() = this.hits(loremCanaryService.container)
-    val CallStats.loremRegularHits: Int
-        get() = this.hits(loremRegularService.container)
-    val CallStats.ipsumRegularHits: Int
-        get() = this.hits(ipsumRegularService.container)
+    override fun envoy() = envoy
```

Comment:
I think we could move EnvoyExtension and ConsulExtension to base interface if we use [Lifecycle PER_CLASS](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/TestInstance.Lifecycle.html#PER_CLASS)